### ISSUE TITLE
Float precision issue

### ISSUE THREAD

**Issue by ailzhang** [taichi-dev/taichi#5059]
Issue from https://forum.taichi.graphics/t/field-float/2612

```
import taichi as ti
ti.init(arch = ti.cpu, print_ir=True)

x= ti.Vector.field(1,ti.f64,8)
x.fill(0.3)
print(x[0][0])
assert x[0][0] == 0.3
```
The script above fails as the python value we got from Taichi field is no longer exactly `0.3`. 

And we seem to have this problem in the beginning since the IR looks like 
```
kernel {
  $0 = offloaded struct_for(S1dense) grid_dim=0 block_dim=8 bls=none
  body {
    <i32> $1 = loop $0 index 0
    <*f64> $2 = global ptr [S2place<f64>], index [$1] activate=false
    <f64> $3 = const [0.30000001192092896]
    $4 : global store [$2 <- $3]
  }
}
```
which implies we have the problem occurs when passing the value from python to C++.

**Comment by bobcao3** [comment]
This is fine? Python floats are by default double while our field by default is single. They have different quantization errors:

FP32, 0.3 -> 0.300000011920928955078125
FP64, 0.3 -> 2.99999999999999988897769753748E-1

**Comment by ailzhang** [comment]
@bobcao3 hmmm but here the field is `f64` so I'd assume it's `2.99999999999999988897769753748E-1` instead of `0.300000011920928955078125`? I think we convert it to f32 first somewhere in the middle and that's something we should fix?

**Comment by bobcao3** [comment]
oh interesting... we probably have somewhere in the code accidentally went to float

**Comment by strongoier** [comment]
The literal is given the default fp type `f32`. Things will be fine if you set `default_fp=ti.f64`.

**Comment by ailzhang** [comment]
@strongoier yup but I'm thinking that we should be more permissive in python literal dtypes?

**Comment by strongoier** [comment]
Yep that will make Python users more comfortable.. But I suppose Taichi picks `f32` instead of `f64` as default type mainly to maintain same behavior across backends, because not all backends support `f64`.

**Comment by ailzhang** [comment]
@strongoier aha good point, thanks! +1 on this will improve python user experience and make it more clear, will hack around that see what we can do ;)
